Sources/DarkModeCore/UIView+DarkModeKit.m

```diff
@@ -5,29 +5,26 @@
 
 #import "UIView+DarkModeKit.h"
 #import "DMDynamicColor.h"
+#import "NSObject+DarkModeKit.h"
 
 @import ObjectiveC;
 
 @implementation UIView (DarkModeKit)
 
-static void (*dm_original_setBackgroundColor)(UIView *, SEL, UIColor *);
-
-static void dm_setBackgroundColor(UIView *self, SEL _cmd, UIColor *color) {
-  if ([color isKindOfClass:[DMDynamicColor class]]) {
-    self.dm_dynamicBackgroundColor = (DMDynamicColor *)color;
-  } else {
-    self.dm_dynamicBackgroundColor = nil;
-  }
-  dm_original_setBackgroundColor(self, _cmd, color);
-}
-
-// https://stackoverflow.com/questions/42677534/swizzling-on-properties-that-conform-to-ui-appearance-selector
 + (void)dm_swizzleSetBackgroundColor {
   static dispatch_once_t onceToken;
   dispatch_once(&onceToken, ^{
-    Method method = class_getInstanceMethod(self, @selector(setBackgroundColor:));
-    dm_original_setBackgroundColor = (void *)method_getImplementation(method);
-    method_setImplementation(method, (IMP)dm_setBackgroundColor);
+    SEL selector = @selector(setBackgroundColor:);
+    [self dm_swizzleSelector:selector withBlock:^id _Nonnull(IMPContainer * _Nonnull container) {
+      return ^(UIView *self, UIColor *backgroundColor) {
+        if ([backgroundColor isKindOfClass:[DMDynamicColor class]]) {
+          self.dm_dynamicBackgroundColor = (DMDynamicColor *)backgroundColor;
+        } else {
+          self.dm_dynamicBackgroundColor = nil;
+        }
+        ((void (*)(UIView *, SEL, UIColor *))container.imp)(self, selector, backgroundColor);
+      };
+    }];
```

Comment:
Using runtime methods directly looks fine to me, what benefits we can get by adding a generic swizzling method?

```objc
SEL selector = @selector(setBackgroundColor:);
Method method = class_getInstanceMethod(self, selector);
__block IMP imp = method_setImplementation(method, imp_implementationWithBlock(^(UIView *self, UIColor *color) {
  if ([color isKindOfClass:[DMDynamicColor class]]) {
    self.dm_dynamicBackgroundColor = (DMDynamicColor *)color;
  } else {
    self.dm_dynamicBackgroundColor = nil;
  }
  ((void (*)(UIView *, SEL, UIColor *))imp)(self, selector, color);
}));
```

Reply:
This is not possible with Swift. Swift will complain that "Variable used within its own initial value" if you use imp in the block, so we have to use a generic method in order to use directly in Swift.

An example would be
```
let block = {
    block()
}
```
This won't compile in Swift.

Reply:
We can use `method_getImplementation` to bypass this issue:

```swift
let selector = #selector(setter: tintColor)
let method = class_getMethodImplementation(UIView.self, selector)!
let imp = method_getImplementation(method)
method_setImplementation(method, imp_implementationWithBlock({ (self: UIView, tintColor: UIColor) -> Void in
  self.dm_dynamicTintColor = tintColor as? DynamicColor
  let oldIMP = unsafeBitCast(imp, to: (@convention(c) (UIView, Selector, UIColor) -> Void).self)
  oldIMP(self, selector, tintColor)
} as @convention(block) (UIView, UIColor) -> Void))
```

And I think we should move code to objc anyway (for swizzlings, simplicity, prevent optimizations by Swift compiler...), we can discuss this in the future.

Reply:
I replaced the calls with direct swizzling, the generic ones does not really reduce the boilerplate codes. Also we can remove the NSObject+DarkModeKit.h plus .m so we don't expose the swizzling method to the users